### PR TITLE
Strict function variance fixes

### PR DESCRIPTION
**Pull request by sandersn**
This PR updates popular packages to work with the upcoming `"strictFunctionTypes": true` option in the upcoming Typescript 2.6 (Microsoft/TypeScript#18654). When this flag is on, the compiler determines whether type parameters should be related covariantly, contravariantly or invariantly and _enforces_ this except for methods. (Methods are an exception because they are so commonly incorrect.)

The fixes fall into 4 categories:

1. Fix bugs in tests found by the errors.
2. Fix definitions.
3. Add type assertions in tests to silence the errors.
4. Hack definitions to allow unsafe, but intended, patterns.

Because methods are exempt from this check, the type `(a?: number) => object` can be rewritten as `{ bivarianceHack(a?: number) => object }['bivarianceHack']`. The second type is identical in almost every way, except that it is related bivariantly &mdash; unsafely. I only used this hack to fix backbone and react, where I believe the pattern is intended even if it's unsafe.

### CODE REVIEW

**Review by johnnyreilly — APPROVED** [review]

**Review by alloy — APPROVED** [review]

File: types/bootbox/index.d.ts [hunk 1]
```diff
@@ -7,9 +7,9 @@
 /// <reference types="jquery" />
 
 /** Bootbox options shared by all modal types */
-interface BootboxBaseOptions {
+interface BootboxBaseOptions<T = any> {
 	title?: string | Element;
-	callback?: (result: boolean | string) => any;
+        callback?: (result: T) => any;
```
Comment (DanielRosenwasser) [inline types/bootbox/index.d.ts]: Nit: fix the indentation here.

File: types/ramda/index.d.ts [hunk 2]
```diff
@@ -225,9 +225,9 @@ declare namespace R {
         /**
          * Returns a new list containing the contents of the given list, followed by the given element.
          */
-        append<T, U>(el: U, list: T[]): Array<(T & U)>;
-        append<U>(el: U): <T>(list: T[]) => Array<(T & U)>;
         append<U>(el: U): <T>(list: T[]) => Array<(T & U)>;
+        append<T, U>(el: U, list: T[]): Array<(T & U)>;
```
Comment (DanielRosenwasser) [inline types/ramda/index.d.ts]: Why is this not `Array<T | U>`? But I guess it was already written this way?

wat

Comment (sandersn) [reply types/ramda/index.d.ts]: convenience!

File: types/bootbox/index.d.ts (hunk 1 above)
Comment (sandersn) [reply types/bootbox/index.d.ts]: Huh, I didn't think anybody used tabs in DT. Fixed.

**Comment by sandersn** [comment]
Note that Travis only gets half way through tests before timing out. I assume that's because I change node and react and jquery. I ran the tests locally and found a couple of failures that I fixed.

**Pull request merged**